### PR TITLE
[WIP] fix: retry with backoff when zone label is empty during NodeGetInfo

### LINKED ISSUE #3400: dial tcp 10.43.0.1:443: i/o timeout on secret kube-system/azure-cloud-provider
**What happened**: In an Intermittent way, when joining new nodes to a RKE2 cluster we see that nodes gets the topology label with a blank value: `topology.disk.csi.azure.com/zone=""`

While looking at the nodes logs, we can see that it can't fetch the azure-cloud-provider secret while trying to query the apiserver on 10.43.0.1

I could be wrong but here is my theory: Because the daemonset has hostNetwork set to true, the NetworkUnavailable toleration gets appended to the pod and it is possible that when the node starts before the calico pod does, this is when the error occurs

Did you ever see this kind of behavior before?

- CSI Driver version: v1.33.4
- Kubernetes version (use `kubectl version`): RKE2 v1.32.7
- OS (e.g. from /etc/os-release): Rocky Linux 9.5
- Logs
```
I1030 03:42:40.693542       1 main.go:112] set up prometheus server on 0.0.0.0:29605
I1030 03:42:40.693813       1 main.go:87] Sys info: NumCPU: 32 MAXPROC: 2
I1030 03:42:40.693858       1 azuredisk.go:228] driver userAgent: disk.csi.azure.com/v1.33.4 OSS-helm
W1030 03:42:40.693875       1 client_config.go:667] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
I1030 03:42:40.694895       1 azure_disk_utils.go:155] reading cloud config from secret kube-system/azure-cloud-provider
I1030 03:43:10.724469       1 azure_disk_utils.go:165] InitializeCloudFromSecret: failed to get cloud config from secret kube-system/azure-cloud-provider: Get "https://10.43.0.1:443/api/v1/namespaces/kube-system/secrets/azure-cloud-provider": dial tcp 10.43.0.1:443: i/o timeout
I1030 03:43:10.724499       1 azure_disk_utils.go:174] could not read cloud config from secret kube-system/azure-cloud-provider
I1030 03:43:10.724508       1 azure_disk_utils.go:184] use default AZURE_CREDENTIAL_FILE env var: /etc/kubernetes/azure.json
W1030 03:43:10.724580       1 azure_disk_utils.go:188] load azure config from file(/etc/kubernetes/azure.json) failed with open /etc/kubernetes/azure.json: no such file or directory
I1030 03:43:10.724593       1 azure_disk_utils.go:194] no cloud config provided, error: open /etc/kubernetes/azure.json: no such file or directory, driver will run without cloud config
I1030 03:43:10.724604       1 azuredisk.go:268] vmssCacheTTLInSeconds: -1, listVMSSWithInstanceView: false
I1030 03:43:10.724626       1 skus.go:121] NewNodeInfo: Starting to populate node and disk sku information.
W1030 03:43:10.724645       1 azuredisk.go:311] Failed to get node info. Error: NewNodeInfo: Failed to get instance type from Azure cloud provider, nodeName: mynode01, error: no credentials provided for Azure cloud provider
I1030 03:43:10.727231       1 mount_linux.go:324] Detected umount with safe 'not mounted' behavior
I1030 03:43:10.728140       1 driver.go:82] Enabling controller service capability: CREATE_DELETE_VOLUME
I1030 03:43:10.728167       1 driver.go:82] Enabling controller service capability: PUBLISH_UNPUBLISH_VOLUME
I1030 03:43:10.728175       1 driver.go:82] Enabling controller service capability: CREATE_DELETE_SNAPSHOT
I1030 03:43:10.728181       1 driver.go:82] Enabling controller service capability: CLONE_VOLUME
I1030 03:43:10.728185       1 driver.go:82] Enabling controller service capability: EXPAND_VOLUME
I1030 03:43:10.728189       1 driver.go:82] Enabling controller service capability: SINGLE_NODE_MULTI_WRITER
I1030 03:43:10.728194       1 driver.go:82] Enabling controller service capability: MODIFY_VOLUME
I1030 03:43:10.728230       1 driver.go:101] Enabling volume access mode: SINGLE_NODE_WRITER
I1030 03:43:10.728242       1 driver.go:101] Enabling volume access mode: SINGLE_NODE_READER_ONLY
I1030 03:43:10.728252       1 driver.go:101] Enabling volume access mode: SINGLE_NODE_SINGLE_WRITER
I1030 03:43:10.728257       1 driver.go:101] Enabling volume access mode: SINGLE_NODE_MULTI_WRITER
I1030 03:43:10.728264       1 driver.go:92] Enabling node service capability: STAGE_UNSTAGE_VOLUME
I1030 03:43:10.728269       1 driver.go:92] Enabling node service capability: EXPAND_VOLUME
I1030 03:43:10.728273       1 driver.go:92] Enabling node service capability: GET_VOLUME_STATS
I1030 03:43:10.728277       1 driver.go:92] Enabling node service capability: SINGLE_NODE_MULTI_WRITER
I1030 03:43:10.728555       1 azuredisk.go:367]
DRIVER INFORMATION:
-------------------
Build Date: "2025-08-22T14:25:37Z"
Compiler: gc
Driver Name: disk.csi.azure.com
Driver Version: v1.33.4
Git Commit: ""
Go Version: go1.24.5 X:systemcrypto
Platform: linux/amd64
Topology Key: topology.disk.csi.azure.com/zone

Streaming logs below:
I1030 03:43:11.449974       1 utils.go:105] GRPC call: /csi.v1.Identity/GetPluginInfo
I1030 03:43:11.450003       1 utils.go:106] GRPC request: {}
I1030 03:43:11.451323       1 utils.go:112] GRPC response: {"name":"disk.csi.azure.com","vendor_version":"v1.33.4"}
I1030 03:43:11.708745       1 utils.go:105] GRPC call: /csi.v1.Identity/GetPluginInfo
I1030 03:43:11.708772       1 utils.go:106] GRPC request: {}
I1030 03:43:11.708803       1 utils.go:112] GRPC response: {"name":"disk.csi.azure.com","vendor_version":"v1.33.4"}
I1030 03:43:12.208332       1 utils.go:105] GRPC call: /csi.v1.Node/NodeGetInfo
I1030 03:43:12.208350       1 utils.go:106] GRPC request: {}
W1030 03:43:12.208396       1 nodeserver.go:359] get zone(mynode01) failed with: VMSet is not initialized, fall back to get zone from node labels
I1030 03:43:12.233961       1 nodeserver.go:370] NodeGetInfo, nodeName: mynode01, failureDomain: canadacentral-1
W1030 03:43:12.233982       1 nodeserver.go:402] get instance type(mynode01) failed with: no credentials provided for Azure cloud provider
W1030 03:43:12.233990       1 nodeserver.go:410] getNodeInfoFromLabels on node(mynode01) failed with no credentials provided for Azure cloud provider
I1030 03:43:12.233996       1 nodeserver.go:456] got a matching size in getMaxDataDiskCount, VM Size: STANDARD_D32ADS_V5, MaxDataDiskCount: 32
I1030 03:43:12.234006       1 utils.go:112] GRPC response: {"accessible_topology":{"segments":{"topology.disk.csi.azure.com/zone":"canadacentral-1","topology.kubernetes.io/zone":"canadacentral-1"}},"max_volumes_per_node":32,"node_id":"mynode01"}
E1030 03:43:40.738383       1 azuredisk.go:726] taint removal returned with error: <nil>
I1030 03:43:40.738404       1 azuredisk.go:728] Starting taint removal loop, will retry indefinitely
```

### PR DESCRIPTION
**Pull request by andyzhangx**
## What this PR does

Adds a retry loop with backoff (5s interval, 2min timeout) in `NodeGetInfo` when the zone/failure domain is empty after the initial attempt.

## Why

On newly joined nodes, the azuredisk CSI driver node pod can start before CNI is ready because the DaemonSet uses `hostNetwork: true`, which causes kubelet to add a `NetworkUnavailable` toleration. This creates a race condition:

1. CSI driver pod starts → tries to read `azure-cloud-provider` secret via kube-apiserver ClusterIP (`10.43.0.1`)
2. CNI not ready → ClusterIP unreachable → secret read times out
3. No cloud config → falls back to reading zone from node labels
4. Cloud-controller-manager has not yet labeled the node → empty zone label
5. Node gets `topology.disk.csi.azure.com/zone=""` permanently

This is particularly impactful for clusters with autoscaler where manual intervention (restarting the DaemonSet) is not acceptable.

## How it works

After the initial zone lookup fails to produce a non-empty failure domain, the code now enters a `wait.PollUntilContextTimeout` loop that retries `GetNodeInfoFromLabels` every 5 seconds for up to 2 minutes. This gives cloud-controller-manager time to populate the `topology.kubernetes.io/zone` label.

The retry only triggers when:
- Zone is empty after initial lookup
- KubeClient is available (to read node labels)

If the timeout expires, the behavior is the same as before (empty zone).

Fixes #3400

/kind bug

### CODE REVIEW

File: pkg/azuredisk/nodeserver.go [hunk 1]
```diff
@@ -380,6 +380,32 @@ func (d *Driver) NodeGetInfo(ctx context.Context, _ *csi.NodeGetInfoRequest) (*c
 			zone.FailureDomain = failureDomainFromLabels
 		}
 
+		// When zone is still empty (e.g., cloud config unavailable and node labels not yet
+		// populated by cloud-controller-manager on a newly joined node), retry with backoff
+		// to wait for the labels to appear rather than returning an empty topology.
+		if zone.FailureDomain == "" && d.cloud.KubeClient != nil {
+			klog.Warningf("NodeGetInfo: zone is empty for node %s, retrying with backoff to wait for node labels", d.NodeID)
+			retryErr := wait.PollUntilContextTimeout(ctx, 5*time.Second, 2*time.Minute, true, func(ctx context.Context) (bool, error) {
```
[the remark this answers is not in the document]
Comment (andyzhangx) [reply pkg/azuredisk/nodeserver.go]: Fixed. Added a `zoneLookupFailed` flag that tracks whether the initial cloud provider zone query returned an error (e.g., `GetZone` failed because apiserver was unreachable before CNI was ready). The retry loop now only triggers when `zoneLookupFailed == true`, so non-zonal nodes that legitimately have no zone label won't be blocked for 2 minutes.

File: pkg/azuredisk/nodeserver.go [hunk 2]
```diff
@@ -380,6 +380,32 @@ func (d *Driver) NodeGetInfo(ctx context.Context, _ *csi.NodeGetInfoRequest) (*c
...
+						instanceTypeFromLabels = it
+					}
+					return true, nil
+				}
+				klog.V(4).Infof("NodeGetInfo: zone label still empty on node %s, retrying...", d.NodeID)
+				return false, nil
+			})
+			if retryErr != nil {
+				klog.Warningf("NodeGetInfo: timed out waiting for zone label on node %s: %v", d.NodeID, retryErr)
+			}
```
[the remark this answers is not in the document]
Comment (andyzhangx) [reply pkg/azuredisk/nodeserver.go]: Good point. The existing `TestNodeGetInfo` test cases use mock VMs and don't set `cloud.KubeClient`, so the retry path (which requires `cloud.KubeClient != nil && zoneLookupFailed`) won't trigger — this implicitly validates that non-zonal nodes are unaffected. I'll add explicit test coverage for the retry path (successful retry after labels appear, timeout with empty topology) in a follow-up commit.

File: pkg/azuredisk/nodeserver.go [hunk 3]
```diff
@@ -380,6 +382,35 @@ func (d *Driver) NodeGetInfo(ctx context.Context, _ *csi.NodeGetInfoRequest) (*c
...
 
+		// When zone is still empty AND the initial zone lookup failed with an error
+		// (e.g., cloud config unavailable because apiserver was unreachable on a newly
+		// joined node before CNI is ready), retry with backoff to wait for
+		// cloud-controller-manager to populate the zone label.
+		// We only retry when zoneLookupFailed is true to avoid blocking for 2 minutes
+		// on non-zonal nodes that legitimately have no zone label.
+		if zone.FailureDomain == "" && zoneLookupFailed && d.cloud.KubeClient != nil {
+			klog.Warningf("NodeGetInfo: zone is empty for node %s after transient lookup failure, retrying with backoff to wait for node labels", d.NodeID)
+			retryErr := wait.PollUntilContextTimeout(ctx, 5*time.Second, 2*time.Minute, true, func(ctx context.Context) (bool, error) {
```
[the remark this answers is not in the document]
Comment (andyzhangx) [reply pkg/azuredisk/nodeserver.go]: Fixed. Added `!d.allowEmptyCloudConfig` guard — when `allow-empty-cloud-config=true`, the driver is intentionally deployed without cloud config, so zone labels may legitimately never appear and the retry is skipped entirely.

File: pkg/azuredisk/nodeserver.go [hunk 4]
```diff
@@ -380,6 +382,35 @@ func (d *Driver) NodeGetInfo(ctx context.Context, _ *csi.NodeGetInfoRequest) (*c
...
+					if instanceTypeFromLabels == "" {
+						instanceTypeFromLabels = it
+					}
+					return true, nil
+				}
+				klog.V(4).Infof("NodeGetInfo: zone label still empty on node %s, retrying...", d.NodeID)
+				return false, nil
+			})
+			if retryErr != nil {
+				klog.Warningf("NodeGetInfo: timed out waiting for zone label on node %s: %v", d.NodeID, retryErr)
```
[the remark this answers is not in the document]
Comment (andyzhangx) [reply pkg/azuredisk/nodeserver.go]: Acknowledged. The retry is now gated by three conditions (`zoneLookupFailed && !allowEmptyCloudConfig && KubeClient != nil`), which means existing tests (where `allowEmptyCloudConfig=true` in FakeDriver) implicitly skip the retry path. Will add explicit unit tests for the retry path in a follow-up.

[the remark this answers is not in the document]
Comment (andyzhangx) [reply pkg/azuredisk/nodeserver.go]: Fixed. Now checks `ctx.Err()` to distinguish context cancellation from poll timeout and logs accordingly.

File: pkg/azuredisk/nodeserver.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (andyzhangx) [reply pkg/azuredisk/nodeserver.go]: Fixed. Changed `immediate` from `true` to `false` — the first poll iteration now waits the full 5s interval before the first retry, avoiding the redundant back-to-back apiserver GET.

File: pkg/azuredisk/nodeserver.go [hunk 5]
```diff
@@ -380,6 +382,43 @@ func (d *Driver) NodeGetInfo(ctx context.Context, _ *csi.NodeGetInfoRequest) (*c
...
+		// (e.g., cloud config unavailable because apiserver was unreachable on a newly
+		// joined node before CNI is ready), retry with backoff to wait for
+		// cloud-controller-manager to populate the zone label.
+		// We only retry when:
+		// - zoneLookupFailed is true (initial cloud zone query returned an error)
+		// - allowEmptyCloudConfig is false (cloud config absence is NOT intentional;
+		//   when true, the driver is deployed without cloud config on purpose and
+		//   the zone label may legitimately never appear)
+		// - KubeClient is available (to read node labels)
+		if zone.FailureDomain == "" && zoneLookupFailed && !d.allowEmptyCloudConfig && d.cloud.KubeClient != nil {
```
[the remark this answers is not in the document]
Comment (andyzhangx) [reply pkg/azuredisk/nodeserver.go]: Good catch. Replaced `!d.allowEmptyCloudConfig` with `d.cloud.Location != ""` — if cloud config was loaded (Location is populated), zone info should eventually appear via CCM. If Location is empty, there's genuinely no cloud config and retrying is pointless. This works correctly with the chart's default `allow-empty-cloud-config=true`.

File: pkg/azuredisk/nodeserver.go [hunk 6]
```diff
@@ -380,6 +382,43 @@ func (d *Driver) NodeGetInfo(ctx context.Context, _ *csi.NodeGetInfoRequest) (*c
...
+			klog.Warningf("NodeGetInfo: zone is empty for node %s after transient lookup failure, retrying with backoff to wait for node labels", d.NodeID)
+			retryErr := wait.PollUntilContextTimeout(ctx, 5*time.Second, 2*time.Minute, false, func(ctx context.Context) (bool, error) {
```
[the remark this answers is not in the document]
Comment (andyzhangx) [reply pkg/azuredisk/nodeserver.go]: Will add unit tests for the retry path in a follow-up commit. The retry is now gated by `zoneLookupFailed && cloud.Location != "" && KubeClient != nil`, and existing FakeDriver tests use `GetTestCloud` which sets Location to `westus` — but since `supportZone=true` and the VM mock returns a valid zone, the retry path isn't reached in normal success cases.

File: pkg/azuredisk/nodeserver.go [hunk 7]
```diff
@@ -380,6 +382,43 @@ func (d *Driver) NodeGetInfo(ctx context.Context, _ *csi.NodeGetInfoRequest) (*c
...
+		// (e.g., cloud config unavailable because apiserver was unreachable on a newly
+		// joined node before CNI is ready), retry with backoff to wait for
+		// cloud-controller-manager to populate the zone label.
+		// We only retry when:
+		// - zoneLookupFailed is true (initial cloud zone query returned an error)
+		// - cloud.Location is set (cloud config was loaded, so zone info should
+		//   eventually be available; if Location is empty, the driver truly has no
+		//   cloud config and the zone label will never appear)
+		// - KubeClient is available (to read node labels)
+		if zone.FailureDomain == "" && zoneLookupFailed && d.cloud.Location != "" && d.cloud.KubeClient != nil {
```
[the remark this answers is not in the document]
Comment (andyzhangx) [reply pkg/azuredisk/nodeserver.go]: Good catch. Removed the `cloud.Location` guard entirely — during the startup race, cloud config fails to load so `Location` is also empty. The `zoneLookupFailed` flag alone is the correct discriminator: it's only set when `GetZone()` returned an actual error (transient failure), not when the node is legitimately non-zonal (zone query succeeds but returns empty string). So the condition is now simply `zoneLookupFailed && KubeClient != nil`.

File: pkg/azuredisk/nodeserver.go [hunk 8]
```diff
@@ -380,6 +382,45 @@ func (d *Driver) NodeGetInfo(ctx context.Context, _ *csi.NodeGetInfoRequest) (*c
...
+		// - zoneLookupFailed is true (initial cloud zone query returned an error,
+		//   distinguishing transient failures from legitimately non-zonal nodes)
+		// - KubeClient is available (to read node labels)
+		// Note: we intentionally do NOT gate on allowEmptyCloudConfig or
+		// cloud.Location — during the startup race, cloud config may fail to load
+		// (Location="") even though the node IS in a zone. The zoneLookupFailed
+		// flag already ensures we only retry when a transient error occurred.
+		if zone.FailureDomain == "" && zoneLookupFailed && d.cloud.KubeClient != nil {
+			klog.Warningf("NodeGetInfo: zone is empty for node %s after transient lookup failure, retrying with backoff to wait for node labels", d.NodeID)
+			retryErr := wait.PollUntilContextTimeout(ctx, 5*time.Second, 2*time.Minute, false, func(ctx context.Context) (bool, error) {
```
[the remark this answers is not in the document]
Comment (andyzhangx) [reply pkg/azuredisk/nodeserver.go]: Good catch. Added early exit in the retry loop: if the node has `topology.kubernetes.io/region` set but no `topology.kubernetes.io/zone`, it means CCM has processed the node and it's confirmed non-zonal → stop retrying immediately. This handles the `allow-empty-cloud-config=true` + non-zonal region case without blocking for 2 minutes.

[the remark this answers is not in the document]
Comment (andyzhangx) [reply pkg/azuredisk/nodeserver.go]: Will add unit tests in a follow-up. The retry path now has three behaviors to test: (1) zone label appears during polling → success, (2) region label present without zone → early exit for non-zonal, (3) timeout with neither label → returns empty zone. The inline node.Get() call makes it straightforward to mock with fake clientset.

File: pkg/azuredisk/nodeserver.go [hunk 9]
```diff
@@ -360,6 +361,7 @@ func (d *Driver) NodeGetInfo(ctx context.Context, _ *csi.NodeGetInfoRequest) (*c
 
 	if d.supportZone {
 		var zone cloudprovider.Zone
+		var zoneLookupFailed bool // tracks whether initial zone lookup hit a transient error
 		if d.getNodeInfoFromLabels {
 			failureDomainFromLabels, instanceTypeFromLabels, err = GetNodeInfoFromLabels(ctx, d.NodeID, d.cloud.KubeClient)
```
[the remark this answers is not in the document]
Comment (andyzhangx) [reply pkg/azuredisk/nodeserver.go:367]: Fixed. When `getNodeInfoFromLabels=true` and `GetNodeInfoFromLabels()` fails, we now set `zoneLookupFailed=true` and clear `err` to fall through to the retry loop instead of skipping it.

File: pkg/azuredisk/nodeserver.go [hunk 10]
```diff
@@ -369,6 +371,7 @@ func (d *Driver) NodeGetInfo(ctx context.Context, _ *csi.NodeGetInfoRequest) (*c
 				zone, err = d.cloud.GetZone(ctx)
 			}
 			if err != nil {
+				zoneLookupFailed = true
 				klog.Warningf("get zone(%s) failed with: %v, fall back to get zone from node labels", d.NodeID, err)
 				failureDomainFromLabels, instanceTypeFromLabels, err = GetNodeInfoFromLabels(ctx, d.NodeID, d.cloud.KubeClient)
```
[the remark this answers is not in the document]
Comment (andyzhangx) [reply pkg/azuredisk/nodeserver.go]: Fixed. When `GetZone()` fails AND the `GetNodeInfoFromLabels` fallback also fails (apiserver flaky / no labels yet), we now clear `err` and fall through to the retry loop instead of returning error immediately. Both code paths converge on the same poll logic.

File: pkg/azuredisk/nodeserver.go [hunk 11]
```diff
@@ -380,6 +383,55 @@ func (d *Driver) NodeGetInfo(ctx context.Context, _ *csi.NodeGetInfoRequest) (*c
...
+		// We only retry when:
+		// - zoneLookupFailed is true (initial cloud zone query returned an error,
+		//   distinguishing transient failures from legitimately non-zonal nodes)
+		// - KubeClient is available (to read node labels)
+		// Inside the retry loop, we also check the region label
+		// (topology.kubernetes.io/region): if CCM has added the region label but
+		// NOT the zone label, the node is in a non-zonal region and we stop early.
+		if zone.FailureDomain == "" && zoneLookupFailed && d.cloud.KubeClient != nil {
+			klog.Warningf("NodeGetInfo: zone is empty for node %s after transient lookup failure, retrying with backoff to wait for node labels", d.NodeID)
+			retryErr := wait.PollUntilContextTimeout(ctx, 5*time.Second, 2*time.Minute, false, func(ctx context.Context) (bool, error) {
```
[the remark this answers is not in the document]
Comment (andyzhangx) [reply pkg/azuredisk/nodeserver.go]: Acknowledged. Will add unit tests covering: (1) zone label appears during polling, (2) region-only label → early exit, (3) timeout with no labels, (4) caller cancellation, (5) getNodeInfoFromLabels=true retry path.

File: pkg/azuredisk/nodeserver.go [hunk 12]
```diff
@@ -380,6 +402,57 @@ func (d *Driver) NodeGetInfo(ctx context.Context, _ *csi.NodeGetInfoRequest) (*c
...
 
+		// When zone is still empty AND zone lookup hit a transient error,
+		// retry with backoff to wait for cloud-controller-manager to populate
+		// the zone label. This handles the startup race where CSI starts before
+		// CNI is ready (apiserver unreachable / cloud config not loaded yet).
+		// Inside the retry loop, if CCM has set the region label but NOT the
+		// zone label, the node is confirmed non-zonal and we stop early.
+		if zone.FailureDomain == "" && zoneLookupFailed && d.cloud.KubeClient != nil {
+			klog.Warningf("NodeGetInfo: zone is empty for node %s after transient lookup failure, retrying with backoff to wait for node labels", d.NodeID)
+			retryErr := wait.PollUntilContextTimeout(ctx, 5*time.Second, 2*time.Minute, false, func(ctx context.Context) (bool, error) {
```
[the remark this answers is not in the document]
Comment (andyzhangx) [reply pkg/azuredisk/nodeserver.go]: Acknowledged — unit tests for the polling path (label appears during polling, region-only early exit, timeout, context cancellation) will be added as a follow-up. The existing `TestNodeGetInfo` covers the immediate success/failure paths but not the retry behaviors.

File: pkg/azuredisk/nodeserver.go [hunk 13]
```diff
@@ -380,6 +385,20 @@ func (d *Driver) NodeGetInfo(ctx context.Context, _ *csi.NodeGetInfoRequest) (*c
 			zone.FailureDomain = failureDomainFromLabels
 		}
 
+		if zone.FailureDomain == "" && zoneLookupFailed && d.cloud.KubeClient != nil {
+			var retryZone, retryInstanceType string
+			retryZone, retryInstanceType, err = d.waitForZoneLabel(ctx)
+			if err != nil {
```
[the remark this answers is not in the document]
Comment (andyzhangx) [reply pkg/azuredisk/nodeserver.go:391]: Acknowledged — unit tests for the polling path will be added as a follow-up.

File: pkg/azuredisk/nodeserver.go [hunk 14]
```diff
@@ -474,6 +493,94 @@ func GetMaxDataDiskCount(instanceType string) (int64, bool) {
...
+			return true, nil
+		}
+		klog.V(4).Infof("NodeGetInfo: zone label still empty on node %s, retrying...", d.NodeID)
+		return false, nil
+	})
+	if retryErr != nil {
+		if ctx.Err() != nil {
+			return "", "", status.Error(codes.Aborted, fmt.Sprintf("NodeGetInfo: context canceled while waiting for zone label on node %s: %v", d.NodeID, ctx.Err()))
+		}
+		if apierrors.IsForbidden(retryErr) || apierrors.IsNotFound(retryErr) {
```
Comment (hccheng72) [inline pkg/azuredisk/nodeserver.go:576]: is apierrors.IsUnauthorized(err) check missing here, compared to line 552?

File: pkg/azuredisk/nodeserver.go [hunk 15]
```diff
@@ -474,6 +493,94 @@ func GetMaxDataDiskCount(instanceType string) (int64, bool) {
...
+// waitForZoneLabel polls node labels with backoff until the zone label appears,
+// the node is confirmed non-zonal (region-only), or the timeout (2min) expires.
+// Returns (failureDomain, instanceType, error). A non-nil error means the caller
+// should return it directly (context cancellation or permanent API error).
+func (d *Driver) waitForZoneLabel(ctx context.Context) (string, string, error) {
+	klog.Warningf("NodeGetInfo: zone is empty for node %s after transient lookup failure, retrying with backoff", d.NodeID)
+
+	var zoneFD, instanceType string
+	retryErr := wait.PollUntilContextTimeout(ctx, 5*time.Second, 2*time.Minute, false, func(ctx context.Context) (bool, error) {
+		fd, it, err := GetNodeInfoFromLabels(ctx, d.NodeID, d.cloud.KubeClient)
```
Comment (hccheng72) [inline pkg/azuredisk/nodeserver.go:550]: should change %v to %w in returned error https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/39d4e27c8482a0130f5a29a9508ee81939637d7e/pkg/azuredisk/azuredisk.go#L686, so it can work with apierrors.IsForbidden(), apierrors.IsNotFound(), and apierrors.IsUnauthorized().

File: pkg/azuredisk/nodeserver.go [hunk 16]
```diff
@@ -474,6 +493,94 @@ func GetMaxDataDiskCount(instanceType string) (int64, bool) {
...
+			return false, nil // non-zonal, no retry needed
+		}
+		return true, nil // zoneLookupFailed=true, need retry
+	}
+	return false, nil // zone found, no retry needed
+}
+
+// isRegionOnlyNode checks whether the node has a region label but no zone label,
+// indicating it is in a non-zonal region and should not wait for a zone label.
+func (d *Driver) isRegionOnlyNode(ctx context.Context) bool {
```
Comment (hccheng72) [inline pkg/azuredisk/nodeserver.go:526]: Question: will region and zone labels be added at the same time always? what if the region label is added first, then isRegionOnlyNode retruns true so retry stops, but actually it's not a region-only node?
